zerver/views/message_fetch.py

```diff
@@ -106,19 +106,40 @@ def get_messages_backend(
     *,
     anchor_val: Annotated[str | None, ApiParamConfig("anchor")] = None,
     include_anchor: Json[bool] = True,
-    num_before: Json[NonNegativeInt],
-    num_after: Json[NonNegativeInt],
+    num_before: Json[NonNegativeInt] = 0,
+    num_after: Json[NonNegativeInt] = 0,
     narrow: Json[list[NarrowParameter] | None] = None,
     use_first_unread_anchor_val: Annotated[
         Json[bool], ApiParamConfig("use_first_unread_anchor")
     ] = False,
     client_gravatar: Json[bool] = True,
     apply_markdown: Json[bool] = True,
+    message_ids_to_fetch: Annotated[
+        Json[list[NonNegativeInt] | None], ApiParamConfig("message_ids")
+    ] = None,
 ) -> HttpResponse:
+    if message_ids_to_fetch is None:
+        message_ids_to_fetch = []
+    # User requested a list of message ids to fetch.
+    is_message_ids_pre_specified = len(message_ids_to_fetch) > 0
+    # User has to either provide message_ids or both num_before and num_after.
+    if (
+        num_before or num_after or anchor_val is not None or use_first_unread_anchor_val
+    ) and is_message_ids_pre_specified:
+        raise JsonableError(
+            _("`message_ids` are mutually exclusive with `num_before`, `num_after` and `anchor`")
+        )
+    elif is_message_ids_pre_specified:
+        include_anchor = False
+
+    anchor = None
+    if not is_message_ids_pre_specified:
+        anchor = parse_anchor_value(anchor_val, use_first_unread_anchor_val)
+
     realm = get_valid_realm_from_request(request)
-    anchor = parse_anchor_value(anchor_val, use_first_unread_anchor_val)
     narrow = update_narrow_terms_containing_with_operator(realm, maybe_user_profile, narrow)
-    if num_before + num_after > MAX_MESSAGES_PER_FETCH:
+
+    if num_before + num_after + len(message_ids_to_fetch) > MAX_MESSAGES_PER_FETCH:
```

Comment:
Cool, this is a nice way to do it. Make sure you have a test covering this.

Reply:
No longer using this after the naming changes.